### ISSUE TITLE
Unable to build on Linux

### ISSUE THREAD

**Issue by Nihlus** [SteveGilham/altcover#3]
I'm currently attempting to build the latest Master branch on Linux (Mint 18.3). I have the Core SDK installed, as well as Mono with the FSharp tooling.

Everything works out up until the fake setup runs, at which point some form of version mismatch occurs and the restore operation fails.

```
jarl@jax ~/Programming/altcover $ dotnet fake run ./Build/setup.fsx
Could not find AppDataFolder, try to set the APPDATA environment variable. Using '/home/jarl/Programming/altcover/.paket' instead.
Could not find LocalApplicationData folder, try to set the 'LocalAppData' environment variable. Using '/home/jarl/Programming/altcover/.paket' instead
/usr/bin/mono --version
/usr/bin/dotnet  restore  ./Build/dotnet-nuget.csproj --packages "./packages"
  Restore completed in 40.56 ms for /home/jarl/Programming/altcover/Build/dotnet-nuget.csproj.
/usr/bin/mono --debug "/home/jarl/Programming/altcover/packages/nuget.commandline/4.5.1/tools/NuGet.exe" "restore" "/home/jarl/Programming/altcover/AltCover.sln" "-OutputDirectory" "/home/jarl/Programming/altcover/packages"
MSBuild auto-detection: using msbuild version '14.0' from '/usr/lib/mono/xbuild/14.0/bin'.
MSBUILD : warning MSB4010: The "*.tasks" files could not be successfully loaded from their expected location "/usr/lib/mono/xbuild/15.0/bin". Default tasks will not be available. Could not find a part of the path '/usr/lib/mono/xbuild/15.0/bin'. [/tmp/NuGetScratch/lielhzwp.zn4.nugetinputs.targets]
/tmp/NuGetScratch/b8k937yz.q15.nugetrestore.targets(219,5): error MSB4036: The "RemoveDuplicates" task was not found. Check the following: 1.) The name of the task in the project file is the same as the name of the task class. 2.) The task class is "public" and implements the Microsoft.Build.Framework.ITask interface. 3.) The task is correctly declared with <UsingTask> in the project file, or in the *.tasks files located in the "/usr/lib/mono/xbuild/15.0/bin" directory. [/tmp/NuGetScratch/lielhzwp.zn4.nugetinputs.targets]

WARNING: Error reading msbuild project information, ensure that your input solution or project file is valid. NETCore and UAP projects will be skipped, only packages.config files will be restored.
All packages listed in packages.config are already installed.
```
A `fake.sh` script is nonetheless generated. When ran, another error (assuming it is related to the above issue) prevents compilation.

```
jarl@jax ~/Programming/altcover $ chmod +x fake.sh 
jarl@jax ~/Programming/altcover $ ./fake.sh 
FsiEvaluationException:

Error: 
	
	/home/jarl/Programming/altcover/Build/actions.fsx(7,11): error FS0039: The namespace 'Core' is not defined.

```
I'm using Mono 5.10.0.118, and dotnet 2.1.3.

**Comment by SteveGilham** [comment]
I had this problem getting the travis-ci build to work -- `dotnet restore` doesn't seem to create a destination directory; instead you have to give it one that already exists to work with.  I got the build working by adding a `mkdir .paket` to the `.travis.yml` file.

If you create the '/home/jarl/Programming/altcover/.paket' does that unblock the build?  If it does, then I can upgrade the "You may need to create the directory .paket" advice to an explicit pre-build step.

**Comment by SteveGilham** [comment]
I've just pushed a "works for me" resolution [acd4459](https://github.com/SteveGilham/altcover/commit/acd445960a7b2e920f62a0a84c9dca7d8ac872b0) that should be zero effort on your part if it does work.

Let me know if that does resolve matters.

**Comment by Nihlus** [comment]
I did have to create the `.paket` folder, yes, but that was before this error. I'm afraid the latest commit exhibits the same problem.

**Comment by SteveGilham** [comment]
Now it's clear that it's not packages failing to arrive, which was my original suspicion, I think I know what's happening.

The wrapper `Build/build.fsx` which is supposed to save the manual effort of updating the script every time there's a new package version is faced with more than one version of FAKE and is picking up the wrong one.  That'll be an all-platform issue, then.

**Comment by SteveGilham** [comment]
Commit [5d5473b](https://github.com/SteveGilham/altcover/commit/5d5473b3b971bdef2dc6c66d6d17078bee276154) should be more resilient to multiple incompatible packages in place.  Does that fix the issue for you?

**Comment by Nihlus** [comment]
That did something - I'm getting further now, however, the original msbuild-related error remains. Now, the build fails due to an inability to resolve the FSharp.Core assembly, and consequently, base library types:


![image](https://user-images.githubusercontent.com/2340428/36349250-4d0cf8a0-1482-11e8-9312-74e7129a28ca.png)

**Comment by SteveGilham** [comment]
So, we've gone from missing the Fake.Core namespaces (because the script would have been referencing FAKE 4.x), to not finding the F# library at a later stage.  

I'll have to think on this one.

**Comment by SteveGilham** [comment]
There are both old and new format versions in the bundle, so I can provide
support back to .net 2.0 (feature parity with OpenCover) as well as going
beyond what that supports.

The problem is with the "am I building on Mono" detection in the classic
framework projects.  If you hang fire for an hour or so, I should have
something into master to try.

**Comment by Nihlus** [comment]
You can still target .NET 2.0 while using the new format and new tooling - it makes nuget a lot smoother, especially in cases like this.

**Comment by Nihlus** [comment]
On top of that, what is requiring the project itself to target .NET 2.0 for it to be able to introspect .NET 2.0 projects?

**Comment by SteveGilham** [comment]
While the main executable can do what it likes, the recorder assembly is
injected into the system under test.  The instrumented assemblies make
calls into its `Visit` API to record when code points are executed, so it
has to be compatible.

**Comment by Nihlus** [comment]
Aha, got it.

**Comment by SteveGilham** [comment]
Give https://github.com/SteveGilham/altcover/commit/c3381bddc22e842ca579fd5a8b2c647edbed607e in master a try.  Amongst other changes, it's cleaned up some of the rather ramshackle "get it to build on travis-ci" parts of the build system, so should, I hope, not now be "only on travis-ci".

**Comment by Nihlus** [comment]
That did it! I'm still seeing unresolved reference warnings about FSharp.Core, but it does mostly compile. The BuildMonoSamples step fails due to this.

**Comment by SteveGilham** [comment]
If you can shoot me a build log (pipe everything to a file), I'll take a look at what's going on.

**Comment by Nihlus** [comment]
Certainly, here you go:
[buildlog.txt](https://github.com/SteveGilham/altcover/files/1739764/buildlog.txt)

**Comment by SteveGilham** [comment]
I'm surprised quite how long the build staggered on, seemingly without a
referenced assembly.

Rather than attempt subtlety, I have switched to the nuget package for
framework builds too (.net 2.0 on Windows aside) in the latest master at
https://github.com/SteveGilham/altcover/commit/b146fb8e159e721de0494b3500ff6722d4627dad
.

Note that in the interim I've also pushed the final phase of the transition
to `dotnet fake`, so the `./fake.sh` build launch is replaced by `dotnet
fake run ./Build/build.fsx`

**Comment by Nihlus** [comment]
The latest commit to master now builds as expected :D Now, the target `JustUnitTest` fails with another peculiar issue, but it doesn't seem to affect the final build. 

https://hastebin.com/omitoqibig.vbs

**Comment by SteveGilham** [comment]
I've seen that particular error intermittently myself.  As you can see from the stack trace, it's NUnit failing to correctly talk to itself across some internal boundary. It may be related to another intermittent NUnit error I've also seen where a TCP link has clearly failed.  In each case, the way forward is to try the build again -- it hasn't yet irked me enough to dig into that code and raise an issue.

Thank you for this report and the assistance you have rendered to help me resolve it.

**Comment by Nihlus** [comment]
Fantastic :) I'm glad to have helped!
